Microsoft.Toolkit.Uwp.UI/Triggers/IsEqualStateTrigger.cs

```diff
@@ -94,6 +94,11 @@ private static object ConvertToEnum(Type enumType, object value)
         {
             try
             {
+                if (value is string str)
```

Comment:
ToObject is only allowed if the type is either 'SByte, Int16, Int32, Int64, Byte, UInt16, UInt32, or UInt64.'.  See https://docs.microsoft.com/en-us/dotnet/api/system.enum.toobject?view=net-5.0#System_Enum_ToObject_System_Type_System_Object_

So I think it's better if we're really clear on which types we convert, so a throw never would happen. Something along the lines of:
```cs
private static object ConvertToEnum(Type enumType, object value)
{
    try
    {
        if (value is string str && Enum.IsDefined(enumType, str))
        {
            return Enum.Parse(enumType, str);
        }
        if((value is SByte || value is Int16 || value is Int32 || value is Int64 || value is Byte || value is UInt16 || value is UInt32 || value is  UInt64) && Enum.IsDefined(enumType, value))
          return  Enum.ToObject(enumType, value);
    }
    catch
    {
    }
    return null;
}
```

Reply:
I don't think such code bloat looks good. Generally we should focus on common scenarios.
Declaring string in xaml is much easier than others. And also, user shouldn't use two objects that looks no way to be compatible.

If both string parsing and number conversion fails, the conversion will probably fail.


Or, in another way, if the calls are guarded well, the try can be removed.

Reply:
@dotMorten I've refined the parsing logic more:
- First, use `object.Equals` instead of `==`. ReferenceEquals can't handle value types.
  - This technically fixes value types with convertType=false
- For strings, use `Enum.TryParse` directly to reduce a call. This also handles new cases of flags combination.
- Check typecode before calling `Enum.IsDefined` and `Enum.ToObject`. This should prevent throw totally.

Reply:
Well, I think the call to `IsDefined` can be removed in case of flags combination.

Reply:
The old code seems return false for `(Flags1.A | Flags1.B, Flags1.A | Flags1.B)`.

Reply:
@Sergio0694 any insights here?

Reply:
@huoyaoyuan @dotMorten was the logic updated after this comment, where are we at with this discussion?

Reply:
@michael-hawker I don't see anything wrong with the solution proposed above of explicitly testing the various types to ensure `ToObject` can be called, though I'd recommend doing that with `GetType()` instead. The JIT has an issue right now where the `is T` pattern isn't properly optimized even when `T` is a sealed type or a value type (I opened an issue about that [here](https://github.com/dotnet/runtime/issues/36649)). For that, I'd recommend using `GetType() == typeof(T)` instead, which is a JIT intrinsic and compiles down to just a type token load and a comparison with the method table pointer.

As in, using the pattern gives us this, currently:

```asm
; return obj is int;
TestType(System.Object)
    L0000: test rcx, rcx
    L0003: je short L0016
    L0005: mov rax, 0x7ff9134fb258
    L000f: cmp [rcx], rax
    L0012: je short L0016
    L0014: xor ecx, ecx
    L0016: test rcx, rcx
    L0019: setne al
    L001c: movzx eax, al
    L001f: ret
```

Whereas using the `GetType() == typeof(T)` intrinsics produces much better codegen:

```asm
; return obj.GetType() == typeof(int);
TestType(System.Object)
    L0000: mov rax, 0x7ff9134fb258
    L000a: cmp [rcx], rax
    L000d: sete al
    L0010: movzx eax, al
    L0013: ret
```

So if we really wanted the best possible solution here, I'd suggest:
- Doing an explicit `null` check beforehand, and fallback (needed for next point)
- Doing a bunch of `obj.GetType() == typeof(int)` for all the supported types for `ToObject()`.
- Then also removing `try/catch` as well, since we should be handling all possible cases anyway.

Of course, reasonably speaking, the cost of everything else going on in a situation where this converter would be used (eg. a layout pass) would likely absolutely dwarf anything going on in this method anyway, but still, it's at least fun to do what we can to optimize things here. Especially for IDE scenarios with an attached debugger though, removing the chance of an exception being thrown will definitely produce the biggest user experience improvement 🙂

Reply:
I don't think it's necessary or possible to maximize performance here, because we use boxing and so on anyway. Use `GetTypeCode` gives best semantic and compatibility.

Reply:
I agree that these performance improvements would likely not be noticeable given the rest of the overhead (as I've mentioned in my comment too), but that said I thought I'd share more info anyway since Michael asked, and also because I don't see anything wrong with still trying to make the code as best we can in whatever context we're in. As in, those two versions are functionally the same, with just a few characters of difference, but one is faster - so we might as well pick the latter just because 😄

Other than that, I do agree that the main win here is just to tweak the code to avoid relying on exception flow and to just avoid the `try/catch` block entirely, that'll definitelly be the most noticeable change for consumers, and I think adding more code to prevent that (like the individual type checks @dotMorten illustrated) is the right thing to do 👍

Reply:
> because I don't see anything wrong with still trying to make the code as best we can in whatever context we're in. As in, those two versions are functionally the same, with just a few characters of difference, but one is faster

It will get the code much more complex with really little improvement comparing to the scenarios where this pattern get used.

First, the values are already boxed, so there's no way to do boxing elimination like pattern in generic `(int)(object)valueOfT`.

Then, there must also be a check for enum underlying check, which is [not effective](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC2A3utmbjgJYB2w2AsgBQAqAngA4Cm2n1ArlDZcAlKXIk05KdkoAzbI14ChnAHQBVagBNOAJwA2rGgHM4SFdgC8l7MA6cA9rMY1gw0ZOlkJXr3ADs2EgA3GJeAL5h0gHYCKGeZJFo4UA==) even used individual type check. *This is literally the biggest reason from me.*

Reply:
> "First, the values are already boxed, so there's no way to do boxing elimination like pattern in generic (int)(object)valueOfT."

I think we might be talking about two different things here.
I was only comparing two approaches to check the type of an object (regardless of whether it's a boxed value type or something else). I never proposed a way to avoid the boxing here, as I agree it's not possible in this context 🙂

These are all marginal points anyway, the main thing is to remove the exception, which we all agree on 😄

Reply:
Well, I was considering about behavior compatibility. Previously `IsDefined` will throw for a number with type other than defined type. But anyway, we are refining the behavior to be better now. It reasonable to accept `int` for `byte` based enum. I'm going to adjust for this.

Reply:
> But anyway, we are refining the behavior to be better now. It reasonable to accept int for byte based enum.

Definitely agree with that, especially in a UI-focused API.
From my experience not many devs know that enums can even have a different backing type in the first place, and in general that's a feature I'd expect to be used more by lower-level devs, rather than frontend devs. Besides, if the converter exposes the target value as `object` and you pass an integer from XAML, it'll be boxed as an `int` anyway, so I agree that giving the API some more internal wiggle room to still work fine in these cases would be nice to have 👍